### PR TITLE
Support PostgresDB config override from NooBaa CR

### LINKED ISSUE #793: postgres config map is overridden by the operator after updating max_connections (or any other)
### Environment info
- NooBaa Operator Version: **master-20211104**
- Platform: **OpenShift 4.8**

### Actual behavior
1. Updated the noobaa-postgres-config CM with max_connections=600
2. The operator overrides the CM back to defaults.
3. Here - https://github.com/noobaa/noobaa-operator/blob/3fbeede64ab2f5e775345c9090375639bc26dd8d/pkg/system/phase2_creating.go#L822-L827

### Expected behavior
1. We want to allow updating the max connections and possibly other configurations without the operator overriding it.

### Steps to reproduce
1. See above.

### PR DESCRIPTION
**Pull request by utkarsh-pro**
Signed-off-by: Utkarsh Srivastava <srivastavautkarsh8097@gmail.com>

### Explain the changes
- This PR allows users to provide postgresql database overrides via NooBaa CR. To achieve so, a new field in the NooBaa CRD had to be added.

### Issues: Fixed #xxx
This PR potentially fixes #793

### Testing Instructions:
1. Deploy operator and operator-catalogue image generated from this PR in a Kubernetes cluster.
2. Run `kubectl edit noobaas.noobaa.io noobaa -n <namespace>`
3. Under `spec` add the following (or any postgresql config):

    ```yaml
    ...
    dbConf: |
      max_connections = 600
    ...
    ```
4. Save the changes
5. Check the db config map to confirm the overrides were indeed applied by running `kubectl get cm noobaa-postgres-config -n <namespace> -o=go-template='{{index .data "noobaa-postgres.conf"}}'`

### Notes For Reviewers
- https://github.com/noobaa/noobaa-operator/issues/793#issuecomment-1033745079 by @dannyzaken mentions this particular approach.
- https://github.com/noobaa/noobaa-operator/issues/793#issuecomment-1033745079 also points out a concern associated with this approach, the concern being: "Will postgresql handle duplicate keys appropriately?". The PR relies on the assumption that it will, following were the reasons to believe so:
    - https://www.postgresql.org/message-id/4F69EB52.7070606@mr-paradox.net thread, where the database users have been relying on this for a while.
    - `postgres.conf`'s `include` directives doesn't appear to be removing duplicate keys: https://github.com/postgres/postgres/blob/master/src/backend/utils/misc/guc-file.l#L735
- PR changes the NooBaa CRD

### Limitations/Caveats
1. Any config passed to `dbConf` is appended to config map without any validation or checks.
2. The field is respected only if the database type is `postgres`, the name of the field doesn't conveys the same.
3. "noobaa-postgres.conf" name is hardcoded and will be the entry in the data that will be modified.
4. The field name `dbConf` could have been `pgDBConf`.

### CODE REVIEW

**Comment by nimrod-becker** [comment]
@dannyzaken @guymguym 
We might want to merge and backport to 5.9 so Spectrum cn use to increase the number of connections.

do you see an issue with this PR?

File: pkg/system/phase2_creating.go [hunk 1]
```diff
@@ -853,6 +853,21 @@ func (r *Reconciler) ReconcileDBConfigMap(cm *corev1.ConfigMap, desiredFunc func
 func (r *Reconciler) SetDesiredPostgresDBConf() error {
 	dbConfigYaml := util.KubeObject(bundle.File_deploy_internal_configmap_postgres_db_yaml).(*corev1.ConfigMap)
 	r.PostgresDBConf.Data = dbConfigYaml.Data
+
+	overrideField := "noobaa-postgres.conf"
+	operator := &nbv1.NooBaa{}
+	if err := r.Client.Get(r.Ctx, types.NamespacedName{Namespace: options.Namespace, Name: "noobaa"}, operator); err != nil {
```
Comment (dannyzaken) [inline pkg/system/phase2_creating.go:859]: the noobaa CR is already fetched at the beginning of the reconciler and you can access it through `r.NooBaa`. no need to get it again

**Review by dannyzaken — COMMENTED** [review]
2 more comments:
* in general, you should squash your commits unless there is a reason to keep them separated
* I think you should add a test for this case

**Comment by dannyzaken** [comment]
> @dannyzaken @guymguym We might want to merge and backport to 5.9 so Spectrum cn use to increase the number of connections.
> 
> do you see an issue with this PR?

I don't see a reason it can't be backported

**Comment by utkarsh-pro** [comment]
> 2 more comments:
> 
> * in general, you should squash your commits unless there is a reason to keep them separated
> * I think you should add a test for this case

@dannyzaken where the test should be added? Should I create a new file under tests?

**Review by guymguym — APPROVED** [review]
@utkarsh-pro Looks good to me. Well done on the testing. See 2 minor code comments.

But more important that these comments - we now have both a `spec.dbConf` string in the noobaa CRD, and also dbconf configmap - both are editable, but there's an operations difference between them - we better explain that difference to users somehow

We should have a section on db configuration explaining these two mechanisms in the [noobaa-crd.md](https://github.com/noobaa/noobaa-operator/blob/master/doc/noobaa-crd.md) doc.

Approving, although I would suggest to include the doc edit (even if brief) in the PR before merging.

**Comment by utkarsh-pro** [comment]
> @utkarsh-pro Looks good to me. Well done on the testing. See 2 minor code comments.
> 
> But more important that these comments - we now have both a `spec.dbConf` string in the noobaa CRD, and also dbconf configmap - both are editable, but there's an operations difference between them - we better explain that difference to users somehow
> 
> We should have a section on db configuration explaining these two mechanisms in the [noobaa-crd.md](https://github.com/noobaa/noobaa-operator/blob/master/doc/noobaa-crd.md) doc.
> 
> Approving, although I would suggest to include the doc edit (even if brief) in the PR before merging.

@guymguym, I am not sure if I understood it clearly so let me try to clarify here.

We have technically 2 ways of updating the PostgreSQL database:
1. Update the `noobaa-postgres-config` config map. Although this config map is editable, we override the user changes during the reconcile process. So users can't really use this config map to override postgres config.
2. Update the `spec.dbConf` in the NooBaa CR. This method does allow to override postgres config.

So is the suggestion to document the above?

File: pkg/system/phase2_creating.go [hunk 2]
```diff
@@ -853,6 +853,13 @@ func (r *Reconciler) ReconcileDBConfigMap(cm *corev1.ConfigMap, desiredFunc func
 func (r *Reconciler) SetDesiredPostgresDBConf() error {
 	dbConfigYaml := util.KubeObject(bundle.File_deploy_internal_configmap_postgres_db_yaml).(*corev1.ConfigMap)
 	r.PostgresDBConf.Data = dbConfigYaml.Data
+
+	overrideField := "noobaa-postgres.conf"
+	operator := r.NooBaa
+	if operator.Spec.DBConf != nil {
+		r.PostgresDBConf.Data[overrideField] = dbConfigYaml.Data[overrideField] + "\n" + *operator.Spec.DBConf
```
[the remark this answers is not in the document]
Comment (utkarsh-pro) [reply pkg/system/phase2_creating.go:862]: I have tried to add a brief comment in the latest commit.

File: pkg/validations/backingstore_validations.go [hunk 3]
```diff
@@ -11,7 +11,7 @@ import (
 
 // const configuration values for the validation checks
 const (
-	MinimumVolumeSize       = 16
+	MinimumVolumeSize       = 16 * 1024 * 1024 * 1024 // 16Gi
```
[the remark this answers is not in the document]
Comment (utkarsh-pro) [reply pkg/validations/backingstore_validations.go]: Ah, I think while squashing commits, I pulled in changes from master. Nevertheless, I have added explicit data type (int64) in the latest commit.

**Comment by guymguym** [comment]
Yes @utkarsh-pro, it is not so clear to users why there is this configmap if its not usable to them, so I suggest to explain how db conf can be set for noobaa.

**Pull request merged**